File: cpp/citor/nqueens.cpp
// Port of cpp/libfork/nqueens.cpp using citor::forkJoinAll.

#include "memusage.hpp"
#include "citor/thread_pool.h"
#include "citor/hints.h"

#include <array>
#include <chrono>
#include <cinttypes>
#include <cstdio>
#include <cstdlib>
#include <thread>

static size_t thread_count = std::thread::hardware_concurrency() / 2;
static const size_t iter_count = 1;

inline constexpr int nqueens_work = 14;

inline constexpr std::array<int, 28> answers = {
  0,       1,         0,          0,          2,           10,     4,
  40,      92,        352,        724,        2'680,       14'200, 73'712,
  365'596, 2'279'184, 14'772'512, 95'815'104, 666'090'624,
};

static void check_answer(int result) {
  if (result != answers[nqueens_work]) {
    std::printf("error: expected %d, got %d\n", answers[nqueens_work], result);
  }
}

template <size_t N>
int nqueens(citor::ThreadPool& pool, int xMax, std::array<char, N> buf) {
  if (xMax == static_cast<int>(N)) {
    return 1;
  }

  std::array<char, N> ys{};
  size_t taskCount = 0;
  for (int y = 0; y < static_cast<int>(N); ++y) {
    char q = static_cast<char>(y);
    bool legal = true;
    for (int x = 0; x < xMax; ++x) {
      char p = buf[x];
      if (q == p || q == p - (xMax - x) || q == p + (xMax - x)) {
        legal = false;
        break;
      }
    }
    if (legal) {
      ys[taskCount++] = static_cast<char>(y);
    }
  }

  if (taskCount == 0) {
    return 0;
  }

  std::array<int, N> values{};
  pool.forkJoinAll<citor::HintsDefaults>(taskCount, [&](size_t i) {
    auto childBuf = buf;
    childBuf[xMax] = ys[i];
    values[i] = nqueens<N>(pool, xMax + 1, childBuf);
  });

  int ret = 0;
  for (size_t i = 0; i < taskCount; ++i) {
    ret += values[i];
  }
  return ret;
}

int main(int argc, char* argv[]) {
  if (argc > 1) {
    thread_count = static_cast<size_t>(atoi(argv[1]));
  }
  std::printf("threads: %zu\n", thread_count);
  citor::ThreadPool pool(thread_count);
  {
    std::array<char, nqueens_work> buf{};
    auto result = nqueens<nqueens_work>(pool, 0, buf); // warmup
    check_answer(result);
  }

  auto startTime = std::chrono::high_resolution_clock::now();

  for (size_t i = 0; i < iter_count; ++i) {
    std::array<char, nqueens_work> buf{};
    auto result = nqueens<nqueens_work>(pool, 0, buf);
    check_answer(result);
    std::printf("output: %d\n", result);
  }

  auto endTime = std::chrono::high_resolution_clock::now();
  auto totalTimeUs =
    std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);
  std::printf("runs:\n");
  std::printf("  - iteration_count: %zu\n", iter_count);
  std::printf("    duration: %" PRIu64 " us\n",
              static_cast<uint64_t>(totalTimeUs.count()));
  std::printf("    max_rss: %ld KiB\n", peak_memory_usage());
  return 0;
}

File: rust/smol/src/bin/nqueens.rs
// Recursive fork-join nqueens.
//
// Canonical (TooManyCooks) implementation:
// ../../../../cpp/TooManyCooks/nqueens.cpp
// which was in turn adapted from libfork's benchmark.
//
// At each column, an iterator over the N rows filters the valid queen
// placements and spawns a child task for each one, collecting the `Task`
// handles (mirroring TMC's `spawn_many` over a task iterator); awaiting them
// in order is the join.

use smol::{future, Executor, Task};
use smol_bench::{default_threads, init_executor, peak_memory_usage, BoxFut};
use std::time::Instant;

const N: usize = 14; // board size / nqueens_work
const ITER_COUNT: usize = 1;

// answers[k] = number of solutions to the k-queens problem.
const ANSWERS: [i32; 19] = [
    0, 1, 0, 0, 2, 10, 4, 40, 92, 352, 724, 2680, 14200, 73712, 365596, 2279184, 14772512,
    95815104, 666090624,
];

fn check_answer(result: i32) {
    if result != ANSWERS[N] {
        println!("error: expected {}, got {}", ANSWERS[N], result);
    }
}

// A queen on row `q` of the next column is valid if it doesn't share a row or
// diagonal with any of the queens already placed in `buf[0..x_max]`.
fn valid_placement(buf: &[i8; N], x_max: usize, q: i32) -> bool {
    (0..x_max).all(|x| {
        let p = buf[x] as i32;
        let d = (x_max - x) as i32;
        q != p && q != p - d && q != p + d
    })
}

fn nqueens(ex: &'static Executor<'static>, x_max: usize, buf: [i8; N]) -> BoxFut<i32> {
    Box::pin(async move {
        if N == x_max {
            return 1;
        }

        let children: Vec<Task<i32>> = (0..N)
            .filter(|&y| valid_placement(&buf, x_max, y as i32))
            .map(|y| {
                let mut next = buf;
                next[x_max] = y as i8;
                ex.spawn(nqueens(ex, x_max + 1, next))
            })
            .collect();

        let mut ret = 0;
        for c in children {
            ret += c.await;
        }
        ret
    })
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let thread_count = if args.len() > 1 {
        args[1].parse().expect("thread count")
    } else {
        default_threads()
    };

    let ex = init_executor(thread_count);
    println!("threads: {}", thread_count);

    {
        let buf = [0i8; N];
        let result = future::block_on(ex.spawn(nqueens(ex, 0, buf))); // warmup
        check_answer(result);
    }

    let start = Instant::now();
    let mut result = 0;
    for _ in 0..ITER_COUNT {
        let buf = [0i8; N];
        result = future::block_on(ex.spawn(nqueens(ex, 0, buf)));
        check_answer(result);
    }
    let dur = start.elapsed();
    println!("output: {}", result);

    println!("runs:");
    println!("  - iteration_count: {}", ITER_COUNT);
    println!("    duration: {} us", dur.as_micros());
    println!("    max_rss: {} KiB", peak_memory_usage());
}
